### ISSUE TITLE
Feature request: uniquely colored nicknames

### ISSUE THREAD

**Issue by ratijas** [barnumbirr/sublime-irc#1]
### Summary

It would be a nice feature (if technically possible) to differentiate users with unique (or at least varying) colors which deterministically based on something like `hash(username)`.

### Version

Sublime Text: Build 4113
IRC Syntax: 1.0.1

### Additional info

The only reason I went out looking for a ST plugin like this, is that I hoped to see dialog messages formatted such that it is clear who's who.

**Comment by UltraInstinct05** [comment]
You can use a special feature available in color schemes called `hashed syntax highlighting`. E.g. 
```json
{
    "name": "IRC",
    "scope": "string.unquoted.username.irc",
    "foreground": ["hsl(200, 60%, 70%)", "hsl(330, 60%, 70%)"],
},
```

The colors can be anything between which ST would interpolate to generate different shades for the target scope text. For more info https://www.sublimetext.com/docs/color_schemes.html#hashed_syntax_highlighting

![image](https://user-images.githubusercontent.com/34304644/132984137-25dac192-4f3d-4221-87a2-cc588a630c7d.png)

This is probably as close as you can get to your request (~~If you want the same usernames to have the same shade, I am afraid that's not possible~~). Posted the same solution on discord https://discord.com/channels/280102180189634562/280157067396775936/886557430790643722

**Comment by ratijas** [comment]
> If you want the same usernames to have the same shade, I am afraid that's not possible

WDYM that's not possible? This is exactly what this feature is all about. Or did I get it wrong? Anyway, how your screenshot would look like with same usernames? @Ultra-Instinct-05

I have to say, official docs are not extremely clear on this topic, and they certainly lack visual examples.

**Comment by ratijas** [comment]
Preferences -> Customize Color Scheme

_Mariana.sublime-color-scheme:_
```json
// Documentation at https://www.sublimetext.com/docs/color_schemes.html
{
    "variables":
    {
    },
    "globals":
    {
    },
    "rules":
    [
        {
            "scope": "string.unquoted.username.irc",
            "foreground": ["hsl(0, 60%, 70%)", "hsl(350, 60%, 70%)"],
        },
    ]
}
```

And suddenly it's become pretty readable:

![hash-colored IRC nicknames](https://user-images.githubusercontent.com/6737986/132984631-e41f87ca-d43d-4e37-abd0-4ce72cf22e28.png)

**Comment by UltraInstinct05** [comment]
Ah, right. I need to read the docs I linked myself better :) I was under the wrong impression each token would receive a different color.
![image](https://user-images.githubusercontent.com/34304644/132984751-ee6a0111-0a5d-491a-839b-30d85e960c2e.png)

**Comment by ratijas** [comment]
Alright, next questions are:

1. How do we ship such color scheme amendment as a part of this plugin?
2. Should we try to standardize `string.unquoted.username` scope and upstream its hash-coloring in themes that are shipped out-of-the-box?

**Comment by UltraInstinct05** [comment]
> How do we ship such color scheme amendment as a part of this plugin?

I am not an author of this package (nor do I use it). It's up to the author to take a call on that. Since this is a syntax package, I personally don't see a need to ship a color scheme and tell the user to use it. It's anyways very easy to customize your custom color scheme to add it.

> Should we try to standardize string.unquoted.username scope and upstream its hash-coloring in themes that are shipped out-of-the-box?

Standardize based on what ? This seems to be a very specific scope assigned in this rather specialized syntax package for highlighting IRC log files. I don't see any need to "ship it out of the box" for default color schemes, although I believe `Celeste` is the only color scheme that supports hashed syntax highlighting out of the box for `string` scopes.

**Comment by ratijas** [comment]
> Since this is a syntax package, I personally don't see a need to ship a color scheme and tell the user to use it.

I mean, if architecture allows override specific scope for all color schemes at once, that would be a good start.

Also, what's this (currently, technically) is — is irrelevant to what it should be in the end. This plugin is for making chat logs easier to read; so if it takes stepping out of being a pure syntax definition file — ok, let it be.

> It's anyways very easy to customize your custom color scheme to add it.

There's a thin line between "very easy" (for whom? did you just assume that users possess any skills?) and "available by default, shipped out of the box".

> Standardize based on what ?

I don't understand, wdym to be "based". But what I mean, is to add it to an official list of scopes and usage recommendations, like "use this for usernames" — IRC is not the only chat system after all, right?

**Comment by UltraInstinct05** [comment]
> Also, what's this (currently, technically) is — is irrelevant to what it should be in the end. This plugin is for making chat logs easier to read; so if it takes stepping out of being a pure syntax definition file — ok, let it be.

What I meant by my earlier statement is, color schemes are a personal thing, syntax is not. By all means, you can ship color schemes, but will a user use it just because it provides hashed syntax highlighting for IRC usernames (& ditch his/her regular color scheme containing his/her favorite colors) where by the same effort they could just add those 5 lines of code to their color schemes and get it ? What makes most sense is for the `README` or a `wiki` in this package to contain this information & present it to the user.

> I don't understand, wdym to be "based". But what I mean, is to add it to an official list of scopes and usage recommendations, like "use this for usernames" — IRC is not the only chat system after all, right?

Exactly my point. `string.unquoted.username` is not a standard scope (I don't know which other chat system logs syntax, if any uses it). Something like `string.quoted.double` is a standardized scope meant to be used in any programming language. It just means additional overhead of adding a non standard scope to the docs that applies only to probably chat system logs (Which personally I don't see any value).

**Comment by ratijas** [comment]
That's why upstreaming/adopting new scope is the only reasonable path forward, as far as see it.

> by the same effort they could just add those 5 lines of code to their color schemes and get it ?

Who are we kidding? No user would ever do that. First and foremost, because they won't even know about it. I never changed a color scheme in ten years that I'm using ST, until today.

> What makes most sense is for the README or a wiki in this package to contain this information & present it to the user.

Agreed. That's a least-effort-but-might-work approach. The [install.txt](https://github.com/barnumbirr/sublime-irc/blob/master/messages/install.txt) file would be a good start.

**Comment by UltraInstinct05** [comment]
> Who are we kidding? No user would ever do that. First and foremost, because they won't even know about it. I never changed a color scheme in ten years that I'm using ST, until today.

I am not kidding anyone. Hashed syntax highlighting is very uncommon as I see it. Yours's is probably the only question I have seen about it in my nearly 2 years in ST discord. Also just because you have never changed a color scheme, doesn't mean others won't change it :)

**Comment by barnumbirr** [comment]
I'll happily merge any PR that adds the mentioned information to the `install.txt` and README. An example with 10-16 colors would be perfect IMHO.

**Comment by ratijas** [comment]
> An example with 10-16 colors would be perfect IMHO.

Not possible to limit/control amount of them — it's always 256 colors interpolated (according to the docs linked above).

**Comment by barnumbirr** [comment]
That's not entirely true, you can make use of CSS colors as follows:

```
// Documentation at https://www.sublimetext.com/docs/color_schemes.html
{
    "variables":
    {
    },
    "globals":
    {
    },
    "rules":
    [
        {
            "name": "IRC username",
            "scope": "string.unquoted.username.irc",
            "foreground": ["var(gold)", "var(red)", "var(skyblue)"]
        },
    ]
}
```

**EDIT**: not just CSS colors, there's a few possible solutions: https://www.sublimetext.com/docs/color_schemes.html#colors

**Comment by barnumbirr** [comment]
Closed in ff7918701b7c41ad3095b9ad4a7d05ff4118fb1f.
